### PR TITLE
Handle custom creatures IDs and exclude them from checks

### PR DESCRIPTION
**Pull request by Bogir**
## Changes Proposed:
-  Created list of custom Creature IDs in worldserver.conf. These IDs will be skipped from some definitions checks

## Tests Performed:

- If you won't to correctly add records for every custom NPC to `gossip_menu` tables - just put all these IDs in this list and server will not distrub you by warnings.
- How @Kitzunu sad, "I dont want to be the one to fix 100 modules. But  the exception is easy". 

## How to Test the Changes:

1. Add your custom IDs to worldserver.conf list `Creatures.CustomIDs` and restart the server.  
2. There should be no messages in the console about those Creatures.
3.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/worldserver/worldserver.conf.dist [hunk 1]
```diff
@@ -1796,6 +1796,16 @@ NpcRegenHPIfTargetIsUnreachable = 1
...
 
+#    Creatures.CustomIDs
+#        Description: The list of custom creatures with gossip dialogues hardcoded in core, 
+#                     divided by "," without spaces.
+#                     It is implied that you do not use for these NPC dialogs data from "gossip_menu" table.
+#                     Server will skip these IDs during the definitions validation process.
+#        Example:     Creatures.CustomIDs = "190010,55005,999991" - Npc for Transmog, Guild-zone and 1v1-arena modules
+#        Default:     ""
+
+Creatures.CustomIDs = ""
```
Comment (Kitzunu) [inline src/server/worldserver/worldserver.conf.dist:1807]:
```suggestion
Creatures.CustomIDs = "190010"
```

**Review by Kitzunu — CHANGES_REQUESTED** [review]
otherwise build is broken :D

**Comment by Nyeriah** [comment]
This is cool, thanks for the PR, I like this idea

**Review by Kitzunu — APPROVED** [review]

File: src/server/worldserver/worldserver.conf.dist [hunk 2]
```diff
@@ -1796,6 +1796,16 @@ NpcRegenHPIfTargetIsUnreachable = 1
 
 NpcRegenHPTimeIfTargetIsUnreachable = 10
 
+#    Creatures.CustomIDs
+#        Description: The list of custom creatures with gossip dialogues hardcoded in core,
```
Comment (Kitzunu) [inline src/server/worldserver/worldserver.conf.dist:1800]:
```suggestion
#        Description: The list of custom creatures with gossip dialogues hardcoded in core,
```

File: src/server/game/Globals/ObjectMgr.cpp [hunk 3]
```diff
@@ -1176,13 +1176,13 @@ void ObjectMgr::CheckCreatureTemplate(CreatureTemplate const* cInfo)
...
-
+    
     for (std::string id; std::getline(stringCreatureIds, id, ',');) // Process each Creature ID in the string, delimited by the comma - ","
     {
         CustomCreatures.push_back(stoul(id));
     }
-
-    for (int32 i = 0; i < CustomCreatures.size(); i++)
+    
+    for (long i = 0; i < CustomCreatures.size(); i++)
```
Comment (Kitzunu) [inline src/server/game/Globals/ObjectMgr.cpp:1185]: use auto

Comment (Bogir) [reply src/server/game/Globals/ObjectMgr.cpp]: very strange comparison requirements  in ubuntu

**Comment by Bogir** [comment]
@Kitzunu need review

**Review by Kitzunu — APPROVED** [review]

File: src/server/game/Globals/ObjectMgr.cpp [hunk 4]
```diff
@@ -1173,11 +1174,20 @@ void ObjectMgr::CheckCreatureTemplate(CreatureTemplate const* cInfo)
     const_cast<CreatureTemplate*>(cInfo)->DamageModifier *= Creature::_GetDamageMod(cInfo->rank);
 
     // Hack for modules
-    switch (cInfo->Entry)
+    std::vector<uint32> CustomCreatures;
+    std::stringstream stringCreatureIds(sConfigMgr->GetOption<std::string>("Creatures.CustomIDs", ""));
```
Comment (Winfidonarleyan) [inline src/server/game/Globals/ObjectMgr.cpp:1178]: No

Comment (Bogir) [reply src/server/game/Globals/ObjectMgr.cpp:1178]: пиши как следует

Comment (Winfidonarleyan) [reply src/server/game/Globals/ObjectMgr.cpp:1178]: `Acore::Tokenize`

Comment (Bogir) [reply src/server/game/Globals/ObjectMgr.cpp:1178]: окей, и дальше какой синтакс?

Comment (Winfidonarleyan) [reply src/server/game/Globals/ObjectMgr.cpp:1178]: Вот пример
https://github.com/azerothcore/azerothcore-wotlk/commit/6d7f58e6ed0f460332cf45b76d879d4a05bb6e24

Comment (Bogir) [reply src/server/game/Globals/ObjectMgr.cpp]: @Winfidonarleyan  review please

File: src/server/game/Globals/ObjectMgr.cpp [hunk 5]
```diff
@@ -1173,11 +1174,20 @@ void ObjectMgr::CheckCreatureTemplate(CreatureTemplate const* cInfo)
...
+    std::string stringCreatureIds(sConfigMgr->GetOption<std::string>("Creatures.CustomIDs", ""));
+    for (std::string_view id : Acore::Tokenize(stringCreatureIds, ',', false))
     {
-        case 190010: // Transmog Module
-          return;
+        uint32 entry = Acore::StringTo<uint32>(id).value_or(0);
+        CustomCreatures.emplace_back(entry);
     }
+
+    for (std::vector<uint32>::const_iterator itr = CustomCreatures.begin(); itr != CustomCreatures.end(); itr++)
```
Comment (Winfidonarleyan) [inline src/server/game/Globals/ObjectMgr.cpp:1185]:
```suggestion
    for (auto const& itr : CustomCreatures)
```

**Comment by Nyeriah** [comment]
Just one thought that occurred me: since we can't set this from inside the module, it means that most modules will still be stuck failing regardless of this, with the exception of the officially supported ones.

**Pull request merged**